compiler/src/dotty/tools/dotc/core/Definitions.scala

```diff
@@ -108,32 +109,28 @@ class Definitions {
    *        def apply(implicit $x0: T0, ..., $x{N_1}: T{N-1}): R
    *      }
    */
-  private def newFunctionNTrait(name: TypeName) = {
+  private def newFunctionNTrait(function: ScalaFunction) = {
     val completer = new LazyType {
       def complete(denot: SymDenotation)(implicit ctx: Context): Unit = {
         val cls = denot.asClass.classSymbol
         val decls = newScope
-        val arity = name.functionArity
-        val argParams =
-          for (i <- List.range(0, arity)) yield
-            enterTypeParam(cls, name ++ "$T" ++ i.toString, Contravariant, decls)
-        val resParam = enterTypeParam(cls, name ++ "$R", Covariant, decls)
+        val tParamTpes = List.tabulate(function.arity + 1) { i =>
+          val variance = if (i < function.arity) Contravariant else Covariant
+          val paramName = tpnme.scala_ ++ "$" ++ function.name ++ "$$" ++ (if (i != function.arity) "T" + (i + 1) else "R")
+          enterTypeParam(cls, paramName, variance, decls).typeRef
+        }
         val (methodType, parentTraits) =
-          if (name.startsWith(tpnme.ImplicitFunction)) {
-            val superTrait =
-              FunctionType(arity).appliedTo(argParams.map(_.typeRef) ::: resParam.typeRef :: Nil)
+          if (function.isImplicit) {
+            val superTrait = function.withoutImplicit.typeRef.appliedTo(tParamTpes)
             (ImplicitMethodType, ctx.normalizeToClassRefs(superTrait :: Nil, cls, decls))
           }
           else (MethodType, Nil)
-        val applyMeth =
-          decls.enter(
-            newMethod(cls, nme.apply,
-              methodType(argParams.map(_.typeRef), resParam.typeRef), Deferred))
-        denot.info =
-          ClassInfo(ScalaPackageClass.thisType, cls, ObjectType :: parentTraits, decls)
+
+        decls.enter(newMethod(cls, nme.apply, methodType(tParamTpes.init, tParamTpes.last), Deferred))
```

Comment:
Now I agree, the change was done on an intermediate version where it saved a considerable amount of code. This is clearly not the case anymore. I will change it back.